src/geom/sides.rs

```diff
@@ -15,6 +15,111 @@ pub struct Sides<T> {
     pub bottom: T,
 }
 
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq, Hash)]
+pub struct Margin {
+    pub sides: Sides<Option<Smart<Rel<Length>>>>,
+    pub two_sided: Option<bool>,
+}
+
+impl Reflect for Margin {
+    fn castable(value: &Value) -> bool {
+        matches!(
+            value,
+            Value::Auto | Value::Dict(_) | Value::Length(_) | Value::Relative(_)
+        )
+    }
+
+    fn describe() -> CastInfo {
+        CastInfo::Type("auto")
+            + CastInfo::Type("dictionary")
+            + CastInfo::Type("length")
+            + CastInfo::Type("relative length")
+    }
+}
+
+impl FromValue for Margin {
+    fn from_value(value: Value) -> StrResult<Self> {
+        match value {
+            Value::Auto => Ok(Self::splat(Some(Value::cast(value)?))),
+            Value::Length(value) => Ok(Self::splat(Some(Smart::Custom(value.into())))),
+            Value::Relative(value) => Ok(Self::splat(Some(Smart::Custom(value)))),
+            Value::Dict(mut dict) => {
+                let mut take = |key| dict.take(key).ok().map(Value::cast).transpose();
+
+                let rest = take("rest")?;
+                let x = take("x")?.or(rest);
+                let y = take("y")?.or(rest);
+                let outside = take("outside")?;
+                let inside = take("inside")?;
+                let mut left = take("left")?;
+                let mut right = take("right")?;
+
+                let implicitly_two_sided = outside.is_some() || inside.is_some();
+                let implicitly_not_two_sided = left.is_some() || right.is_some();
+
+                if implicitly_two_sided && implicitly_not_two_sided {
+                    panic!("Error: Cannot determine if file is two-sided. Use either outside and inside margins or left and right margins, but not both.");
```

Comment:
Don't panic here, as this is not a compiler error, but a user error. Instead, return an `Err` so that the error is nicely displayed to the user:

```suggestion
                    return Err("Cannot determine if the page is two-sided. Use either outside and inside margins or left and right margins, but not both.");
```

Reply:
Done.